### ISSUE TITLE
Issues with pivoting query on singleton Collections

### ISSUE THREAD

**Issue by oliverfleetwood** [orientechnologies/orientdb#5870]
Hi, check the below example:

```SQL
let v1 = create Vertex set label ="v1"
let v2 = create Vertex set sameValue = true, diffValue = 0
let v3 = create Vertex set sameValue = true, diffValue = 1
create Edge E from $v1 To $v2
create Edge E from $v1 To $v3
let failed1 = select 'size() is not working after fetching the first object in a singleton collection.' as description, out('E')[diffValue=0][0].size() as size from $v1
let working1 = select 'size() works on a singleton collection.' as description, out('E')[diffValue=0].size() as size from $v1
let working2 = select 'size() works after fetching the first object in a non-singleton collection.' as description, out('E')[sameValue=true][0].size() as size from $v1
let working3 = select 'size() works after fetching the first object in a singleton collection which has been converted to a list.' as description, list(out('E')[diffValue=0])[0].size() as size from $v1

return [$failed1,$working1,$working2,$working3]
```

We expect size to be 1 in all queries, but as seen in the $failed1 there is a problem when there is just one object in the collection. 

Though it seems to be a limited use case this becomes an issue when you don't know if there are multiple objects in the collection after pivoting or just one. The workaround is to always put the pivoted elements in a list, which I suppose will cause some overhead. 

This issue occurred when we upgraded from orientdb 2.1.9 to 2.1.13.

**Comment by oliverfleetwood** [comment]
Update: my workaround is no good. If you have multiple records in the result set the list will be taken over all returned records (for example, replace ```from $v1``` with ```from V``` in the query above)

**Comment by luigidellaquila** [comment]
Hi @oliverfleetwood 

that is a legacy behavior, when you filter with square brackets and the result is a single value, the collection is unwrapped and the item is returned alone.

Anyway, why are you doing `out('E')[diffValue=0][0].size()`? IMHO it would make more sense to do out('E')[diffValue=0].size()

Thanks

Luigi

**Comment by oliverfleetwood** [comment]
Hi again, so I know my test case is a bit stupid but that's just to show what is going on. We've had some real issues with this when we use it in LET block. I'll come back with a test case in a sec. 

How can this be legacy behaviour though since it came into effect when we upgraded from orientdb 2.1.9 to 2.1.13?

**Comment by oliverfleetwood** [comment]
A more realistic use case, filtering on edges properties: 
 * Assume that we have two fields on our edges, sameValue and diffValue. We want to fetch records by filtering based on edge properties. There can be any number of edges with the different values. 
 * Run the batch script below and then the query, but with `:vertexId` replaced by the id returned from the batch query. Do not merge the two queries into one batch query! Then everything will magically work. I believe this is a potential bug as well.

```SQL
let v1 = create Vertex set label ="v1"
let v2 = create Vertex set label = "v2"
let v3 = create Vertex set label = "v3"
create Edge E from $v1 To $v2 set sameValue = true, diffValue = 1
create Edge E from $v1 To $v3 set sameValue = true, diffValue = 0
return $v1
```

```SQL
let failed = select 'size() does not work' as description, outE()[diffValue=0].size() from :vertexId where outE()[diffValue=0].size() > 0
let working = select 'size() does work' as description, outE()[sameValue=true].size() from :vertexId where outE()[sameValue=true].size() > 0
let workAround = select 'size() does work with workaround' as description, $projection.size() from :vertexId let $projection = list(outE()[diffValue=0]) where $projection.size() > 0
return list($failed,$working, $workAround)
```

**Comment by luigidellaquila** [comment]
Ok, I tried it on 2.1.7 and 2.1.15 and I got your point.

I'm trying to figure out what changed. I'll give you a feedback asap

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @oliverfleetwood 

I just pushed a fix to 2.1.x branch, now I'm porting it to develop

Thanks

Luigi

**Comment by luigidellaquila** [comment]
(btw, you were right, it was a regression)

**Comment by luigidellaquila** [comment]
Ported to develop. Closing

The fix will be available in v 2.1.16 and 2.2 RC1

Thanks

Luigi

**Comment by oliverfleetwood** [comment]
Great thanks!
